jax/experimental/lanczos.py

```diff
@@ -0,0 +1,308 @@
+# Copyright 2020 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Lanczos methods for solving large symmetric eigenvalue problems.
+
+The "basic" Lanczos method has severe numerical stability issues when
+implemented with non-exact arithmetic. "Full reorthogonalization" fixes these
+issues, but requires increases the runtime from O(k) to O(k**2), which quickly
+becomes prohibitively expensive.
+
+Restarted Lanczos methods use restarts to avoid the quadratic scaling. In
+particular, the "thick restart" variant we implement here is reported to be the
+approach of choice for large-scale eigenvalue problems [2].
+
+References:
+[1] Wu, K., Simon, H. D. & Wang, L.-W. Thick-Restart Lanczos Method for
+    Electronic Structure Calculations. Journal of Computational Physics 154,
+    156–173 (1999). https://escholarship.org/content/qt6gk8b245/qt6gk8b245.pdf
+[2] V. Hernandez, J. E. Roman, A. Tomas, and V. Vidal. Krylov-schur methods
+    in SLEPc. Technical report, Universidad Politecnica de Valencia, 2007.
+    http://slepc.upv.es/documentation/reports/str7.pdf
+
+Future improvements:
+- Tests for Hermitian (non-symmetric) matrices (with complex eigenvectors)
+- JVP and VJP rules (needed for automatic differentitation).
+- Support for finding the largest eigenvalue as well.
+- Port to jax.scipy.sparse.linalg.eigsh
+- Preconditioning for finding non-perpheral eigenvalues and solving
+  generalized eigenvalue problems.
+- Consider also implementing LOBPCG (a non-Lanczos method).
+"""
+
+import collections
+from functools import partial
+
+import jax
+from jax import lax
+import jax.numpy as jnp
+
+
+_dot = partial(jnp.dot, precision=lax.Precision.HIGHEST)
+
+
+def _iterative_classical_gram_schmidt(Q, x, iterations=2):
+  """Orthogonalize x against the columns of Q."""
+  # "twice is enough"
+  # http://slepc.upv.es/documentation/reports/str1.pdf
+
+  q = x
+  r = 0
+
+  for _ in range(iterations):
+    h = _dot(Q.T.conj(), q)
+    q = q - _dot(Q, h)
+    r = r + h
+
+  return r, q
+
+
+def _lanczos_restart(A, k, m, Q, alpha, beta):
+  """The inner loop of the restarted Lanzcos method."""
+
+  def body_fun(i, carray):
+    Q, alphas, betas = carray
+    q_prev = Q[:, i]
```

Comment:
Hi! I noticed that Krylov vectors are stored in column major format here. Would it make sense to switch to row-major?  Wondering if for large-dimensional vectors, zero-padding on TPUs could lead to an unnecessarily large memory footprint.